Userland/Libraries/LibTextCodec/Encoder.cpp

```diff
@@ -0,0 +1,589 @@
+/*
+ * Copyright (c) 2024, Ben Jilks <benjyjilks@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/BinarySearch.h>
+#include <AK/Error.h>
+#include <AK/Utf8View.h>
+#include <LibTextCodec/Decoder.h>
+#include <LibTextCodec/Encoder.h>
+#include <LibTextCodec/LookupTables.h>
+
+namespace TextCodec {
+
+namespace {
+UTF8Encoder s_utf8_encoder;
+GB18030Encoder s_gb18030_encoder;
+GB18030Encoder s_gbk_encoder(GB18030Encoder::IsGBK::Yes);
+Big5Encoder s_big5_encoder;
+EUCJPEncoder s_euc_jp_encoder;
+ISO2022JPEncoder s_iso_2022_jp_encoder;
+ShiftJISEncoder s_shift_jis_encoder;
+EUCKREncoder s_euc_kr_encoder;
+}
+
+Optional<Encoder&> encoder_for_exact_name(StringView encoding)
+{
+    if (encoding.equals_ignoring_ascii_case("utf-8"sv))
+        return s_utf8_encoder;
+    if (encoding.equals_ignoring_ascii_case("big5"sv))
+        return s_big5_encoder;
+    if (encoding.equals_ignoring_ascii_case("euc-jp"sv))
+        return s_euc_jp_encoder;
+    if (encoding.equals_ignoring_ascii_case("iso-2022-jp"sv))
+        return s_iso_2022_jp_encoder;
+    if (encoding.equals_ignoring_ascii_case("shift_jis"sv))
+        return s_shift_jis_encoder;
+    if (encoding.equals_ignoring_ascii_case("euc-kr"sv))
+        return s_euc_kr_encoder;
+    if (encoding.equals_ignoring_ascii_case("gb18030"sv))
+        return s_gb18030_encoder;
+    if (encoding.equals_ignoring_ascii_case("gbk"sv))
+        return s_gbk_encoder;
+    dbgln("TextCodec: No encoder implemented for encoding '{}'", encoding);
+    return {};
+}
+
+Optional<Encoder&> encoder_for(StringView label)
+{
+    auto encoding = get_standardized_encoding(label);
+    return encoding.has_value() ? encoder_for_exact_name(encoding.value()) : Optional<Encoder&> {};
+}
+
+// https://encoding.spec.whatwg.org/#utf-8-encoder
+ErrorOr<void> UTF8Encoder::process(Utf8View input, ErrorMode, Function<ErrorOr<void>(u8)> on_byte)
+{
+    for (auto item : input) {
+        // 1. If code point is end-of-queue, return finished.
+
+        // 2. If code point is an ASCII code point, return a byte whose value is code point.
+        if (item < 0x0080) {
+            TRY(on_byte(item));
+            continue;
+        }
+
+        // 3. Set count and offset based on the range code point is in:
+        size_t count = 0;
+        size_t offset = 0;
+        if (item >= 0x0080 && item <= 0x07FF) {
```

Comment:
The first check here is redundant. The same is true for the two `else if`s below.

Reply:
We now implement this in the way skyrising suggested.